### ISSUE TITLE
ASP.Net Core 1.0 compatibility

### ISSUE THREAD

**Issue by davidetaddeucci** [sendgrid/sendgrid-csharp#221]
Any news about the release of Sendgrid C# API libraries compatible with ASP.Net 5.0 (or ASP.Net Core 1.0)?
Just added the reference to the last version of this library on my ASP.Net Core 1.0 using NuGet manager project but there's an error in the dependencies saying that SendGrid API is not compatible with DNX 5.0.

**Comment by saluce65** [comment]
See issue #146.

**Comment by 0xdeafcafe** [comment]
I wrote a small library for covering a limited usage of the v2 mail send endpoint as we needed coverage at my company. It supports DotNet Core, if you want to use it, it's over [here](https://github.com/0xdeafcafe/sendgrid-dotnet). It would be nice if SendGrid supported DotNet Core on the `v3beta` branch however.

**Comment by davidetaddeucci** [comment]
Dear Alex,

thanks a lot for your support.

Unfortunatelt in these days I have opted for an other massive mailing service provider, SparkPost; but I would try also your interesting solution.

Best regards from Tuscany.

Davide

 

 

Da: Alex Forbes-Reed [mailto:notifications@github.com] 
Inviato: giovedì 19 maggio 2016 15:26
A: sendgrid/sendgrid-csharp <sendgrid-csharp@noreply.github.com>
Cc: davidetaddeucci <info@hybrid.it>; Author <author@noreply.github.com>
Oggetto: Re: [sendgrid/sendgrid-csharp] ASP.Net Core 1.0 compatibility (#221)

 

I wrote a small library for covering a limited usage of the v2 mail send endpoint as we needed coverage at my company. It supports DotNet Core, if you want to use it, it's over here <https://github.com/0xdeafcafe/sendgrid-dotnet> . It would be nice if SendGrid supported DotNet Core on the v3beta branch however.

**Comment by thinkingserious** [comment]
@davidetaddeucci this is definitely on our roadmap, see: https://github.com/sendgrid/csharp-http-client/issues/1

@0xdeafcafe Awesome! *Highest of Fives*

I'm going to leave this ticket open to make sure we support .NET Core on the v3beta branch too.

Thanks everyone for your support!

**Comment by RyanONeill1970** [comment]
Can we have a list of action points and get these agreed by the project owners? I can then sign a CLA and work on some bits.

From what I understand, the SendGrid project that handles the API v3 calls only has 3 points that need addressing to turn it into a Portable Class Library that can be used by the different .Net variants.

The Mail project (in the same solution) however has that huge dependency on System.Net.Mail (which will be ported, but they only seem to have started it just a few days back, see https://github.com/dotnet/corefx/issues/1006 ). I suggest splitting the projects so that the API can go fully PCL and be used by .Net core.

If the project owners think this is OK and can split the solution then I can list the points required to convert the API to a PCL based class.

It would mean effectively stopping working on the older Mail project until .Net core catches up with a new implementation of System.Net.Mail.

**Comment by thinkingserious** [comment]
@RyanONeill1970,

Thank you for the feedback!

So far we have two pull requests that are attempting a solution: https://github.com/sendgrid/csharp-http-client/pulls. Have you checked those out? If so, what do you think?

**Comment by RyanONeill1970** [comment]
I was talking about the API3 beta3 branch at
https://github.com/sendgrid/sendgrid-csharp/tree/v3beta , the branch you
pointed to is a lower level one I think (not so strongly typed).
However, those pull requests look good.

They'll need to be re-done though as I'm sure you are aware, a new beta
came out last week which overlapped them.

Looks like you have plenty of help anyway, I'll stand back and watch.

Thanks

**Comment by thinkingserious** [comment]
Yes, the pull requests refer to the HTTP client that the v3beta version of
this library depends upon. I'm thinking if we get it right there, making
the necessary changes in this project should be straight forward.

Thanks for your support!

<http://www.sendgrid.com>
Elmer Thomas
Developer Experience Engineer | Product

**Comment by cleftheris** [comment]
Hi all any updates?

**Comment by thinkingserious** [comment]
@cleftheris 

Thanks for checking in!

We have one signed CLA from @PaitoAnderson, but I was hoping to get one for this pull request from @Gekctek also: https://github.com/sendgrid/csharp-http-client/pull/2 so that we could compare notes and possibly merge solutions. 

That said, we just took this library out of beta yesterday (along with our 6 others) and we will likely be spending a few weeks ironing out bugs before we come back to enhancements such as this one.

**Comment by bdparrish** [comment]
@thinkingserious 

I am following this update closely.  Is there any update for this, so that it can be used with Azure Developer Services?

**Comment by yellowsix** [comment]
For anyone who just needs the basic functionality of sending email via the SendGrid API from a .NET Core application, it's actually pretty easy to use the REST API directly. The C# API in this library was definitely written by someone who is not a .NET developer... it is very disorienting to use with the weird dynamic type stuff and lack of typical naming for async methods and the such. To actually send an email you only need a few lines of code:

```
    // you can/should reuse the http client instance for multiple requests, 
    // then Dispose when done with it. I recommend making a singleton class
    // with this client as a member, and registering it in Startup.cs of your ASP.NET Core app, 
    // just as an example
    var client = new HttpClient();
    client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", "your api key");
    client.BaseAddress = new Uri("https://api.sendgrid.com/v3/");

    // sending JSON to the API is as easy as a single line of code:
    var response = await client.PostAsync("mail/send", new StringContent(json, Encoding.UTF8, "application/json"));
```

And to get the `json` string to send, you can make a little object model on the REST API here:
https://sendgrid.com/docs/API_Reference/Web_API_v3/Mail/index.html

For example, 5 minutes of coding gets you this:

```
        public class SendGridMessage
        {
            public const string TYPE_TEXT = "text";
            public const string TYPE_HTML = "text/html";

            public List<SendGridPersonalization> personalizations { get; set; }
            public SendGridEmail from { get; set; }
            public List<SendGridContent> content { get; set; }

            public SendGridMessage() { }

            public SendGridMessage(SendGridEmail to, string subject, SendGridEmail from, string message, string type = TYPE_HTML)
            {
                personalizations = new List<SendGridPersonalization> { new SendGridPersonalization { to = new List<SendGridEmail> { to }, subject = subject } };
                this.from = from;
                content = new List<SendGridContent> { new SendGridContent(type, message) };
            }
        }

        public class SendGridPersonalization
        {
            public List<SendGridEmail> to { get; set; }
            public string subject { get; set; }
        }

        public class SendGridEmail
        {
            public string email { get; set; }
            public string name { get; set; }

            public SendGridEmail() { }

            public SendGridEmail(string email, string name = null)
            {
                this.email = email;
                this.name = name;
            }
        }

        public class SendGridContent
        {
            public string type { get; set; }
            public string value { get; set; }

            public SendGridContent() { }

            public SendGridContent(string type, string content)
            {
                this.type = type;
                value = content;
            }
        }
```

Then you can use JSON.NET to serialize your object as a string:

```
        public async Task SendEmailAsync(string to, string subject, string fromAddress, string fromName, string message)
        {
            var msg = new SendGridMessage(new SendGridEmail(to), subject, new SendGridEmail(fromAddress, fromName), message);
            try
            {
                string json = JsonConvert.SerializeObject(msg);
                var response = await _httpClient.PostAsync("mail/send", new StringContent(json, Encoding.UTF8, "application/json"));

                // this is just a rough example of handling errors
                if (!response.IsSuccessStatusCode)
                {
                    // see if we can read the response for more information, then log the error
                    string errorJson = await response.Content.ReadAsStringAsync();
                    throw new Exception($"SendGrid indicated failure, code: {response.StatusCode}, reason: {errorJson}");
                }
                
            }
            catch (Exception ex)
            {
                // example using a fictional object "_logger" that can log exceptions for you
                await _logger.LogExceptionAsync(ex);
            }
        }
```

In this example, the HttpClient created above is stored as _httpClient in a class that I have for sending emails. Included is some rough example of logging errors, but not failing on an email that can't be sent. Change as appropriate for your situation.

Hopefully that will help some people who don't want to include this library just to send some simple emails from a .NET Core application.

**Comment by thinkingserious** [comment]
Thank you for the contribution @yellowsix!

**Comment by thinkingserious** [comment]
@yellowsix if you would like some SendGrid swag, please email us at dx@sendgrid.com with your mailing address and T-shirt size. Thanks again!

**Comment by 0xdeafcafe** [comment]
@thinkingserious can I get some new SendGrid swag? ;)

**Comment by thinkingserious** [comment]
@0xdeafcafe of course you can! Just email us with your mailing address and T-shirt size.

**Comment by edrohler** [comment]
@thinkingserious Are you looking for help from the community on this library? From what it sounds like a wrapper needs to be built for the http calls in v3.

**Comment by thinkingserious** [comment]
@edrohler,

You are correct. While we will do what we can to make the improvements internally, any help from the community is appreciated and encouraged.

There are many improvements identified (see our issues and pull requests for a start) for the next iteration of this library, including wrappers executed in the form of helpers.

If you are interested, I'm happy to dig deeper with you.

**Comment by edrohler** [comment]
@thinkingserious I would be happy to help. I use this library for a couple of projects and would like to see it progress.

**Comment by thinkingserious** [comment]
Thanks @edrohler!

As a first step, could you please submit a signed CLA? https://github.com/sendgrid/sendgrid-csharp/blob/master/CONTRIBUTING.md#cla

Thanks!

**Comment by Bartmax** [comment]
I just want to add another way of using sendgrid for sending emails only purpose.

You can use [MailKit](https://github.com/jstedfast/MailKit) which already supports ASP.NET Core and use SMTP settings as described here:
https://sendgrid.com/docs/Classroom/Basics/Email_Infrastructure/recommended_smtp_settings.html
```
var message = new MimeMessage();
message.From.Add(new MailboxAddress("John Doe", fromEmail));
message.ReplyTo.Add(new MailboxAddress("John Doe", fromEmail));
message.To.Add(new MailboxAddress("Bart Calixto", toEmail));
message.Subject ="mail subject";

var msg = "mail message";
message.Body = new TextPart("plain") { Text = msg };
using (var client = new SmtpClient())
{
    await client.ConnectAsync("smtp.sendgrid.net", 587, SecureSocketOptions.Auto);
    await client.AuthenticateAsync("user", "pass");
    await client.SendAsync(message);
    await client.DisconnectAsync(true);
}
```

**Comment by onovotny** [comment]
Ping? What's the status here? It also seems like the use of dynamic on the client makes this API very "undiscoverable". If you're looking for an easy REST wrapper, I would recommend looking at github.com/paulcbetts/refit, as it fits the bill very cleanly. 

Simply define your POCO types and define an interface that is decorated with the API methods and it does the rest. Much more discoverable that way.

**Comment by thinkingserious** [comment]
Hi @onovotny,

Thanks for the up vote, you have helped this issue rise in our queue. 

If the world stood still, it looks like we would get this done in a few weeks. This project can increase in the queue with additional +1's. 

Do you mind opening a new ticket regarding adding POCO types?

**Comment by thinkingserious** [comment]
@onovotny,

Awesome, that one needed some love. I give your +1 to that ticket in our backlog to help it move up the queue and pinged the author to get that CLA on file.

Thanks!

**Comment by onovotny** [comment]
@thinkingserious curious about your CI process? Not sure why #246 failed, but I noticed you're using xbuild on travis with Mono? 

I hate to say it but, xbuild is rarely up to date and is not an ideal way to build/test/run a .NET 4.5 library "accurately." 

Http client issues aside, it appears that the best path to get this client with maximum platform support is to target netstandard 1.1 or 1.2. The best way to do that today is with a portable csproj that targets netstandard -- my [blog post](https://oren.codes/2016/06/23/portable-is-dead-long-live-netstandard/) shows how to do that, but you will need a Windows build agent with VS 2015 Update 3 on it. 

Happy to help/review as needed.

**Comment by thinkingserious** [comment]
@onovotny,

Those tests failed because at that time we were using an external service (stoplight.io) to create a mock SendGrid server and the variable that held the host URL was in an environment variable only accessible to people on the SendGrid team.

We have since updated that process so that the mock SendGrid server now runs locally in Travis.

With regards to your other comments, do you mind opening a separate issue for that. Perhaps "Improve CI Process" or something similar. Your help would be greatly appreciated! Note that currently, all our development is done the VS Community Edition. Our current thoughts are that we don't want to add any dependencies that would require someone to use a paid version of Visual Studio. Of course, I'd be happy to test this with your direction.

**Comment by onovotny** [comment]
Sure, I'll file a seperate issue. VS 2015 Community is fine; it's fully featured and is all that'd be required on a build agent.

That said, you can also use AppVeyor for free as an open source project. A very simple AppVeyor.yml config file is all that's needed.

**Comment by thinkingserious** [comment]
Hello Haris Botić,

We received your email, but your email address did not come through. Could you email us directly at dx@sendgrid.com? Thanks!

**Comment by cpius** [comment]
+1

**Comment by kilnan** [comment]
+1

**Comment by thinkingserious** [comment]
Thanks @cpius and @kilnan! Your votes have been added.

**Comment by harisbotic** [comment]
+1

**Comment by rogeliovidaurri** [comment]
+1

**Comment by thinkingserious** [comment]
Thanks @harisbotic and @rogeliovidaurri! Your votes have been added.

**Comment by emanuel-lundman** [comment]
+1

**Comment by OG84** [comment]
+ 1

**Comment by Vasimovic** [comment]
+1

**Comment by snaegeli** [comment]
+1

**Comment by thinkingserious** [comment]
Hello Everyone,

This is going to be a breaking change, most likely. Could you please lend us a hand and take a look at this: https://github.com/sendgrid/sendgrid-csharp/issues/317

We think your feedback will help make this process much smoother.

**Comment by KhuongNtrd** [comment]
I have new pull request (#320) porting to support dotnet core.

**Comment by szykov** [comment]
+1 folks, please add support for .net core framework. even if you're not targeting cross-platform but want to use VS Code or third party IDE you still need to use .net core instead of classic .net framework 4.x.

only sendgrid client holds me from migrating to .net core. thanks.

**Comment by chassq** [comment]
Just adding our vote for this... Tried to add SendGrid 8.0.4 to a project using "NETStandard.Library": "1.6.0" with no luck. Hope this support is delivered soon. Would be helpful. 

For now we are writing our notifications to a queue and have a worker process using core full framework pick up the notifications and sending them out. Pretty stable work around but would like support in SendGrid for NetStandard lib... Thanks!!!

**Comment by thinkingserious** [comment]
Just FYI.

When we implement this, we are looking to natively support 4.5 and 4.6 simultaneously along side .NET Core.

**Comment by dstep2007** [comment]
+1

**Comment by 0xdeafcafe** [comment]
You should look at the [.NET Standard](https://blogs.msdn.microsoft.com/dotnet/2016/09/26/introducing-net-standard/).

Target the library at the standard, and then have any additional functionality that only works on x platform with compiler framework conditional statements ([like so](https://github.com/XboxChaos/Assembly/commit/6bbdad46c6efe59bf9a5741c2dbaaf3348e573cc#diff-8b632b89c698043403a258abd11bd068R1)). They can be defined in the [project.json](https://github.com/XboxChaos/Assembly/blob/dev/src/Blamite/project.json#L49).

**Comment by tinchou** [comment]
I think @0xdeafcafe is right, but you probably won't even need to use conditional compilation. You'll only need an HTTP client for talking to your server, and that is available everywhere.

Also, I'm thinking for now this is a good approach for me, while you guys implement this: https://github.com/sendgrid/sendgrid-csharp/issues/221#issuecomment-230575041.

**Comment by 0xdeafcafe** [comment]
@tinchou very true. However if there was SMTP support, or using the `MailAddress` model you would need to as they are not in the .NET Standard.

**Comment by tinchou** [comment]
For SMTP you could use a general purpose library (I think they mentioned
MailKit), and this library could provide support for the HTTP api.

Is there any reason this library has to support SMTP?

**Comment by 0xdeafcafe** [comment]
I don't see why it should, and I would personally be against it containing an SMTP option. I was just saying if they were going to, that would be a way to solve the issue.

**Comment by kspearrin** [comment]
Patiently waiting for this update to land.

**Comment by sdaub** [comment]
+1

**Comment by thinkingserious** [comment]
@kspearrin @sdaub thank you for your votes and patience!

**Comment by kmute90** [comment]
+1

**Comment by mattosaurus** [comment]
+1

**Comment by thinkingserious** [comment]
@kmute90 @mattosaurus thanks for the votes! We are getting closer, you can follow along [here](https://github.com/sendgrid/sendgrid-csharp/tree/v9beta).

**Comment by 0xdeafcafe** [comment]
In the mean time, i've updated my dotnet-core library to use the final v3 json mail api.

Repo: https://github.com/0xdeafcafe/sendgrid-dotnet
Nuget: https://www.nuget.org/packages/SendGridClient/2.0.0-preview1
Example:
``` csharp
var key = new ApiKeyConnection("SG.api.key");
var client = new SendGridClient(key);
client.MailClient.SendAsync(new Email
{
    Personalizations = new List<Personalization>
    {
        new Personalization
        {
            To = new List<EmailDetail>
            {
                new EmailDetail
                {
                    Email = "customer@yahoooooo.com",
                    Name = "Customer Name"
                }
            }
        }
    },
    From = new EmailDetail
    {
        Email = "no-reply@company.domain",
        Name = "Company Letter"
    },
    Subject = "sup",
    Content = new List<Content>
    {
        new Content
        {
            Type = "text/html",
            Value = "<h1>yo yo! :)</h1>"
        }
    }
}).Wait();
```

**Comment by Jericho** [comment]
For anyone waiting for a SendGrid library they can use in their .Net Core project, I just want to mention that I released a library called (StrongGrid)[https://www.nuget.org/packages/StrongGrid/] which covers the entire SendGrid v3 API which includes sending emails (of couse!) and also allows you to manage lists and segments, allows you to create custom fields for your lists, allows you to import contacts in bulk, etc.

StrongGrid supports multiple .Net frameworks such as 4.5.2, 4.6, 4,6.1 and 4.6.2 and also .Net standard but most importantly, it's strongly typed which means no more `dynamic`!

**Comment by RyanONeill1970** [comment]
Thank you @Jericho , I was going to move to another provider because of my distrust of the dynamic approach but this solves it for me.

**Comment by mbernier** [comment]
To be fair, we are VERY close to having the new version without dynamic completed. We realized this  was a mistake (my fault) and it is being resolved as quickly as we can.

**Comment by Jericho** [comment]
@mbernier Great to hear that your library is almost ready and that you will fix the `everything must be dynamic` design ;-)

I suspect however that you are only addressing the email aspect of your API (which I believe you cal the "email helper") am I right? As far as I know you are not addressing al the other aspects of your API such as list management, contacts bulk import, creation of api keys, searching for recipients matching criterias, etc. I mention this to point out the StrongGrid addresses all of these aspects.

By the way, I hope you'll remember our February 2016 hour long phone conversation where I gave you the feedback that `dynamic` is not convenient at all and would be rejected by developers. Too bad that I wasn't able to convince you to change your design at that time and that you decided not to accept the PR I submitted last spring to resolve this issue.

**Comment by tinchou** [comment]
Let's not turn this discussion into a war. Open source allows people to choose from whatever library they prefer, and I'm sure both projects will benefit from now on. And if some people would rather wait to have a company-backed library, I respect their decision.

Personally, I really appreciate the effort you put into StrongGrid!

**Comment by fa10** [comment]
But @Jericho StrongGrid doesn't support .net core?

**Comment by mbernier** [comment]
The plan is to completely get rid of dynamic from the entire library as well as support .Net core.

**Comment by fa10** [comment]
@mbernier so no support for .NET Core? Wut

**Comment by Jericho** [comment]
@fa10 StrongGrid is built on top of .Net Standard and therefore compatible with .Net Core

**Comment by mbernier** [comment]
Oh, PHRASING!!! 

The plan is to completely get rid of dynamic from the entire library [AND THEN ADD] ~~as well as~~ support for .Net core.


The plan is to completely get rid of dynamic from the entire library **and then add** support for .Net core.


Sorry about that. My english, despite being my only language, is awful.

**Comment by Jericho** [comment]
@tinchou Thanks for the encouragement. Also, edited my previous comment based on your feedback to make sure it doesn't come across as a "declaration of war".

**Comment by mbernier** [comment]
It should also be noted that we consider @Jericho a good friend and we support and appreciate the work he is doing on StrongGrid. We have been in some AMAZING conversations with him about our own library and his. We share ideas and thoughts, learning from him as we go.

**Comment by threevaluelogic** [comment]
+1

Just wanted to add my support for this! Sendgrid is my chosen email client and I would love to use it in my .net core solutions :).

**Comment by thinkingserious** [comment]
Thanks @threevaluelogic, I've added your vote to our queue.

**Comment by losviko** [comment]
+1!

**Comment by thinkingserious** [comment]
Thanks for the vote @losviko! Added.

**Comment by dwdickens** [comment]
+1

**Comment by garethrampton** [comment]
+1

**Comment by thinkingserious** [comment]
Thanks @dwdickens,

This issue is currently now under development :)

**Comment by kshyju** [comment]
I am waiting for this. 👍 .

**Comment by thinkingserious** [comment]
Hello Everyone,

I have pushed the support for .NET Core to the [v9beta branch](https://github.com/sendgrid/sendgrid-csharp/tree/v9beta). The SendGrid library now targets .Net Standard and no longer has the dynamic dependency. 

First, HUGE thanks to @Jericho, @0xdeafcafe, @Gekctek and @PaitoAnderson for their feedback and code contributions! More HUGE thanks to everyone on this thread for their patience and feedback!

In the Solution (in the v9beta branch) you will find two example projects, one using .NET Core and one using .NET 4.6. There are also Unit Tests (more like Integration Tests), but they only work in Visual Studio (I specifically have tested with the 2015 Community Edition) in Windows 10. I also uploaded a hidden Nuget package and tested the following configurations:
- .NET Core in Visual Studio Community Edition 2015 Windows
- .NET 4.5 in Visual Studio Community Edition 2015 Windows
- .NET 4.6 in Visual Studio Community Edition 2015 Windows
- .NET Core in Visual Studio Mac
- .NET 4.5 in Visual Studio for Mac
- .NET 4.6 in Visual Studio Mac 

Unfortunately, I can not get the tests running on Travis CI, any help there would be appreciated.

Before I push the v9beta branch to Nuget as a beta release, there is one more task, the [update of the Mail Helper](https://github.com/sendgrid/sendgrid-csharp/issues/331), which is the next item on our list. Please see that thread for updates. It is likely that I will publish the new interface there for feedback within the next few days.

With Best Regards,

Elmer

**Comment by aKzenT** [comment]
Are there any news on when there will be a beta or non-beta version on Nuget available? Thank you!

**Comment by thinkingserious** [comment]
@aKzenT,

I expect a beta to be released by early next month at the latest. Since our backlog is dynamically prioritized and based on several variables, I can not promise an exact date. That said, this issue is currently our top priority.

**Comment by kierenj** [comment]
@thinkingserious Would you perhaps mind posting again here when it's available please?  I'm subscribed to this thread & don't know where else to monitor!

**Comment by thinkingserious** [comment]
@kierenj,

I'm planning to release a beta to Nuget this week based on the current v9beta branch. I'll post here when that goes live :)

Thanks!

**Comment by thinkingserious** [comment]
Hello everyone,

BIG thanks to all of those that have contributed on this thread, we greatly appreciate your patience and support!

The [v9beta branch](https://github.com/sendgrid/sendgrid-csharp/tree/v9beta) has been pushed to Nuget as a [prerelease](https://www.nuget.org/packages/Sendgrid/9.0.0-beta) for testing. Please give it a spin and let us know what you think :)

The main updates are:
- Removed dynamic dependencies
- Added support for .NET Standard 1.3
- Updated the Mail Helper for better UX/DX

Here is what we think needs to be completed before we exit beta:
- A few weeks of iterative improvements based on your feedback and internal testing
- Updated docs (USAGE, USE_CASES, TROUBLESHOOTING, CONTRIBUTING, /examples and the Helper Example)
- Update the auto-generated portions of the tests to use CamelCase function names
- Move from Travis to Appveyor
- Implement SendSingleEmailToMultipleRecipientsAsync
- Implement SendMultipleEmailsToMultipleRecipientsAsync
- Adopt StyleCop

The remainder of our wants for this version will be added to GitHub as issues.

With Best Regards,

Elmer

**Comment by 0xdeafcafe** [comment]
Awesome, can't wait to have a play around with it. One quick Q:

> Update the auto-generated portions of the tests to use CamelCase function names
  
When you say `CamelCase`, do you mean `camelCase` or `TitleCase`?

**Comment by thinkingserious** [comment]
Hi @0xdeafcafe,

Thanks!

Yes, I mean TitleCase. Then, with StyleCop I hope to avoid such mistakes moving forward. In this particular case it was a matter of me using the wrong function name generator (snake case vs camel case).

**Comment by kierenj** [comment]
Works a treat for my (admittedly simple) use case, thank you :)

**Comment by thinkingserious** [comment]
Thanks for taking the time to provide some feedback @kierenj :)

**Comment by szykov** [comment]
@thinkingserious I just tried and it worked like a charm. Haven't found where I can set an encoding for an email object but it works with Cyrillic out-of-the-box. Also noticed that right now we need to specify **PlainTextContent** and **HtmlContent** at the same time. The first idea was that it isn't right, that there had to be overloaded method for each one, but then I realized that it was done for compatibility reasons if email-client can only work with a plain text only, I believe so, please correct me if I'm wrong.
Anyway, thank you @thinkingserious  and SendGrid team for your wok. :1st_place_medal:

**Comment by thinkingserious** [comment]
Thanks for taking the time to give it a try and provide feedback @szykov! Also, for the kind words :)

While it's not required to have both plain/text and plain/html, it is highly recommended to include both for maximal delivery (including the case you mentioned). So in this case, we are forcing some best practices via the helper example. That said, I don't think the helper deals with the case where you leave one of those out, so I'll get that fixed.

If you build the SendGridMessage helper yourself, you do not need to add both mime types.

**Comment by thinkingserious** [comment]
@szykov,

I've updated the code to make sure the payload is still building correctly if you leave out the text or html: https://github.com/sendgrid/sendgrid-csharp/blob/v9beta/src/SendGrid/Helpers/Mail/SendGridMessage.cs#L1406

Thanks again for the heads up!
